File: JDK11/opentraceability/src/main/java/opentraceability/models/events/aggregationevent`1.java
package opentraceability.models.events;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import opentraceability.interfaces.EPCISVersion;
import opentraceability.interfaces.EventAction;
import opentraceability.interfaces.EventBase;
import opentraceability.interfaces.EventBusinessTransaction;
import opentraceability.interfaces.EventDestination;
import opentraceability.interfaces.EventILMD;
import opentraceability.interfaces.EventLocation;
import opentraceability.interfaces.EventProduct;
import opentraceability.interfaces.EventReadPoint;
import opentraceability.interfaces.EventSource;
import opentraceability.interfaces.IAggregationEvent;
import opentraceability.interfaces.IEventKDE;
import opentraceability.interfaces.IILMDEvent;
import opentraceability.interfaces.OpenTraceabilityArrayAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityAttribute;
import opentraceability.interfaces.OpenTraceabilityObjectAttribute;
import opentraceability.interfaces.OpenTraceabilityProductsAttribute;
import opentraceability.interfaces.OpenTraceabilityProductsAttribute;
import opentraceability.interfaces.OpenTraceabilityProductsAttribute;
import opentraceability.interfaces.OpenTraceabilityProductsListType;
import opentraceability.models.identifiers.EPC;
import opentraceability.utility.attributes.PersistentDisposition;
import opentraceability.utility.attributes.SensorElement;

import java.net.URI;
import java.util.ArrayList;
import java.util.List;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
@SuppressWarnings({"MismatchedQueryAndUpdateOfCollection", "unused", "FieldCanBeLocal", "InnerClassMayBeStatic"})
public class AggregationEvent<T extends EventILMD> extends EventBase implements IAggregationEvent, IILMDEvent<T> {
    @JsonProperty("@type")
    public EventType type = EventType.TransformationEvent;

    @OpenTraceabilityAttribute("", "parentID", 7)
    public EPC parentID = null;

    @OpenTraceabilityProductsAttribute(value = "", version = EPCISVersion.V1, type = EventProductType.Child, order = 21, listType = OpenTraceabilityProductsListType.QuantityList)
    @OpenTraceabilityProductsAttribute(value = "childQuantityList", version = EPCISVersion.V2, type = EventProductType.Child, order = 15, listType = OpenTraceabilityProductsListType.QuantityList)
    @OpenTraceabilityProductsAttribute(value = "childEPCs", version = EPCISVersion.V2, type = EventProductType.Child, order = 8, listType = OpenTraceabilityProductsListType.EPCList, havingAttributes = true)
    public List<EventProduct> children = new ArrayList<>();

    @OpenTraceabilityAttribute("", "action", 9)
    public EventAction action = null;

    @OpenTraceabilityAttribute("", "bizStep", 10)
    public URI businessStep = null;

    @OpenTraceabilityAttribute("", "disposition", 11)
    public URI disposition = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "readPoint", 12)
    public EventReadPoint readPoint = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "bizLocation", 13)
    public EventLocation location = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("bizTransaction")
    @OpenTraceabilityAttribute("", "bizTransactionList", 14)
    public List<EventBusinessTransaction> bizTransactionList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("source")
    @OpenTraceabilityAttribute("", "sourceList", 16, EPCISVersion.V2)
    @OpenTraceabilityAttribute(value = "", v1Order = 22, version = EPCISVersion.V1)
    public List<EventSource> sourceList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("destination")
    @OpenTraceabilityAttribute("", "destinationList", 17, EPCISVersion.V2)
    @OpenTraceabilityAttribute(value = "", v1Order = 23, version = EPCISVersion.V1)
    public List<EventDestination> destinationList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("sensorElement")
    @OpenTraceabilityAttribute("", "sensorElementList", 18, EPCISVersion.V2)
    public List<SensorElement> sensorElementList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "persistentDisposition", 19, EPCISVersion.V2)
    public PersistentDisposition persistentDisposition = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "ilmd", 20, EPCISVersion.V2)
    public T ilmd = null;

    @OpenTraceabilityXmlIgnoreAttribute
    @OpenTraceabilityAttribute("", "type", 0)
    public EventType eventType = EventType.AggregationEvent;

    public List<EventProduct> products = new ArrayList<>();

    public AggregationEvent() { }

    public AggregationEvent(T ilmd) {
        this.ilmd = ilmd;
    }

    public AggregationEvent(EventReadPoint readPoint, EventAction action) {
        this.readPoint = readPoint;
        this.action = action;
    }

    @SuppressWarnings("unchecked")
    public <T extends IEventKDE> T getKDE(Class<T> clazz, String ns, String name) {
        return super.getKDE(clazz, ns, name);
    }

    @SuppressWarnings("unchecked")
    public <T extends IEventKDE> T getKDE(Class<T> clazz) {
        return super.getKDE(clazz);
    }

    public void addProduct(EventProduct product) {
        switch (product.Type) {
            case Parent:
                if (product.Quantity != null) throw new Exception("Parents do not support quantity.");
                this.parentID = product.EPC;
                break;

            case Child:
                this.children.add(product);
                break;
            default:
                throw new Exception("Aggregation event only supports children and parents.");
        }
    }

    public EventILMD grabILMD() {
        return ilmd;
    }
}

File: Java/OpenTraceability/src/main/kotlin/opentraceability/models/events/aggregationevent`1.kt
package opentraceability.models.events

import com.fasterxml.jackson.annotation.*
import opentraceability.interfaces.*
import opentraceability.models.identifiers.EPC
import opentraceability.utility.attributes.*
import java.net.URI
import kotlin.collections.*

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
@Suppress("LocalVariableName", "PropertyName")
class AggregationEvent<T : EventILMD>(
    @JsonProperty("@type")
    val type: EventType = EventType.TransformationEvent,

    @OpenTraceabilityAttribute("", "parentID", 7)
    override var parentID: EPC? = null,

    @OpenTraceabilityProductsAttribute(
        "extension/childQuantityList",
        EPCISVersion.V1,
        EventProductType.Child,
        21,
        OpenTraceabilityProductsListType.QuantityList
    )
    @OpenTraceabilityProductsAttribute("childQuantityList", EPCISVersion.V2, EventProductType.Child, 15, OpenTraceabilityProductsListType.QuantityList)
    @OpenTraceabilityProductsAttribute("childEPCs", EPCISVersion.V2, EventProductType.Child, 8, OpenTraceabilityProductsListType.EPCList,  true)
    var children: MutableList<EventProduct> = mutableListOf(),

    @OpenTraceabilityAttribute("", "action", 9)
    override var action: EventAction? = null,

    @OpenTraceabilityAttribute("", "bizStep", 10)
    override var businessStep: URI? = null,

    @OpenTraceabilityAttribute("", "disposition", 11)
    override var disposition: URI? = null,

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "readPoint", 12)
    override var readPoint: EventReadPoint? = null,

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "bizLocation", 13)
    override var location: EventLocation? = null,

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("bizTransaction")
    @OpenTraceabilityAttribute("", "bizTransactionList", 14)
    override var bizTransactionList: MutableList<EventBusinessTransaction> = mutableListOf(),

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("source")
    @OpenTraceabilityAttribute("", "sourceList", 16, EPCISVersion.V2)
    @OpenTraceabilityAttribute("","baseExtension/sourceList", 22, EPCISVersion.V1)
    override var sourceList: MutableList<EventSource> = mutableListOf(),

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("destination")
    @OpenTraceabilityAttribute("", "destinationList", 17, EPCISVersion.V2)
    @OpenTraceabilityAttribute("","baseExtension/destinationList", 23, EPCISVersion.V1)
    override var destinationList: MutableList<EventDestination> = mutableListOf(),

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("sensorElement")
    @OpenTraceabilityAttribute("", "sensorElementList", 18, EPCISVersion.V2)
    override var sensorElementList: MutableList<SensorElement> = mutableListOf(),

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "persistentDisposition", 19, EPCISVersion.V2)
    override var persistentDisposition: PersistentDisposition? = null,

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "ilmd", 20, EPCISVersion.V2)
    override var ilmd: T? = null
) : EventBase(), IAggregationEvent, IILMDEvent<T> {

    @OpenTraceabilityXmlIgnoreAttribute
    @OpenTraceabilityAttribute("", "type", 0)
    override var eventType: EventType = opentraceability.models.events.EventType.AggregationEvent

    override var products: MutableList<EventProduct> = mutableListOf()
        get() {
            val prods: MutableList<EventProduct> = mutableListOf()
            this.parentID?.let {
                prods.add(EventProduct(it).apply { Type = EventProductType.Parent })
            }
            prods.addAll(this.children)
            return prods
        }

    override fun <T: IEventKDE> getKDE(clazz: Class<T>, ns: String, name: String): T? {
        return super.getKDE<T>(clazz, ns, name)
    }

    override fun <T: IEventKDE> getKDE(clazz: Class<T>): T? {
        return super.getKDE<T>(clazz)
    }

    override fun addProduct(product: EventProduct) {
        when (product.Type) {
            EventProductType.Parent -> {
                if (product.Quantity != null) {
                    throw Exception("Parents do not support quantity.")
                }
                this.parentID = product.EPC
            }

            EventProductType.Child -> {
                this.children.add(product)
            }

            else -> {
                throw Exception("Aggregation event only supports children and parents.")
            }
        }
    }

    override fun grabILMD(): EventILMD? {
        return ilmd
    }
}